src/irr/asset/CGLSLVirtualTexturingBuiltinIncludeLoader.h

```diff
@@ -0,0 +1,368 @@
+	int clippedLoD = positiveLoD ? min(LoD_high,originalMaxFullMip):0;
+
+	// if minification is being performaed then get tail position
+	int levelInTail = LoD_high-clippedLoD;
+	// have to do trilinear only if doing minification AND larger than 1x1 footprint
+	bool haveToDoTrilinear = levelInTail<int(PAGE_SZ_LOG2) && positiveLoD;
+	levelInTail = haveToDoTrilinear ? levelInTail:(positiveLoD ? int(PAGE_SZ_LOG2):0);
+
+	// get the higher resolution mip-map level
+	vec3 hiPhysCoord = irr_glsl_vTextureGrad_helper(formatID,virtualUV,clippedLoD,levelInTail);
+	// get lower if needed (speculative execution, had to move divergent indexing to a single place)
+    vec3 loPhysCoord;
+	// speculative if (haveToDoTrilinear)
+	{
+		// now we have absolute guarantees that both LoD_high and LoD_low are in the valid original mip range
+		bool highNotInLastFull = LoD_high<originalMaxFullMip;
+		clippedLoD = highNotInLastFull ? (clippedLoD+1):clippedLoD;
+		levelInTail = highNotInLastFull ? levelInTail:(levelInTail+1);
+		loPhysCoord = irr_glsl_vTextureGrad_helper(formatID,virtualUV,clippedLoD,levelInTail);
+	}
+
+	vec4 hiMip_retval;
+    vec4 loMip;
+#ifdef IRR_GL_EXT_nonuniform_qualifier
+    hiMip_retval = textureGrad(physicalTileStorageFormatView[nonuniformEXT(formatID)],hiPhysCoord,dOriginalScaledUV[0],dOriginalScaledUV[1]);
+    if (haveToDoTrilinear)
+        loMip = textureGrad(physicalTileStorageFormatView[nonuniformEXT(formatID)],loPhysCoord,dOriginalScaledUV[0],dOriginalScaledUV[1]);
+#else
+    uint64_t outstandingSampleMask = ballotARB(true);
+    // maybe unroll a few times manually
+    while (outstandingSampleMask!=uint64_t(0u))
+    {
+		uvec2 tmp = unpackUint2x32(outstandingSampleMask);
+        uint subgroupFormatID = readInvocationARB(formatID,tmp[1]!=0u ? 32u:findLSB(tmp[0]));
+        bool canSample = subgroupFormatID==formatID; // do I need this? && (outstandingSampleMask&gl_SubGroupEqMaskARB)==gl_SubGroupEqMaskARB;
+        outstandingSampleMask ^= ballotARB(canSample);
+        if (canSample)
+        {
+            hiMip_retval = textureGrad(physicalTileStorageFormatView[subgroupFormatID],hiPhysCoord,dOriginalScaledUV[0],dOriginalScaledUV[1]);
+            if (haveToDoTrilinear)
+                loMip = textureGrad(physicalTileStorageFormatView[subgroupFormatID],loPhysCoord,dOriginalScaledUV[0],dOriginalScaledUV[1]);
+        }
+    }
+#endif
+    if (haveToDoTrilinear)
+	    hiMip_retval = mix(hiMip_retval,loMip,LoD-float(LoD_high));
+    return hiMip_retval;
+}
+
+float irr_glsl_wrapTexCoord(float tc, in uint mode)
+{
+    switch (mode)
+    {
+    case irr_glsl_STextureData_WRAP_REPEAT: tc = fract(tc); break;
+    case irr_glsl_STextureData_WRAP_CLAMP:  tc = clamp(tc, 0.0, 1.0); break;
+    case irr_glsl_STextureData_WRAP_MIRROR: tc = 1.0 - abs(mod(tc,2.0)-1.0); break;
+    default: break;
+    }
+    return tc;
+}
+vec4 irr_glsl_textureVT(in uvec2 _texData, in vec2 uv, in mat2 dUV)
```

Comment:
this needs to be called `irr_glsl_vTextureGrad` and `irr_glsl_vTextureGrad` needs to become something else (maybe `irr_glsl_vTextureGrad_impl`)